schema_automator/generalizers/csv_data_generalizer.py

```diff
@@ -645,6 +645,12 @@ def infer_range(slot: dict, vals: set, types: dict, coerce=True) -> str:
         if all(isfloat(v) for v in nn_vals):
             return 'float'
         if all(is_date(v) for v in nn_vals):
+            if all(
+                not hasattr(parse(str(v)), 'hour') or
```

Comment:
It would be best if you only called `parse` once for each value. You could extract this conditional logic into a function like `def guess_is_datetime(v: datetime) -> bool`, and then refactor this into:

```
parsed_datetimes = [parse(str(v)) for v in nn_vals]
if all(guess_is_datetime(v) for v in parsed_datetimes):
    return "datetime"
return "date"
```

Reply:
I did something slightly different because I realized there is a funny edge case... what if there are some dates and some datetimes? But I can't decide what to do so I'm going to default those to dates and fail it in validation. I don't like raising an exception here and it shouldn't be string.